### PR TITLE
Cors pagination helpers

### PR DESCRIPTION
**Pull request by pdyba**

### CODE REVIEW

File: lbz/resource.py [hunk 1]
```diff
@@ -86,3 +91,86 @@ def pre_request_hook(self):
...
+        "Authorization",
+        "X-Api-Key",
+        "X-Amz-Security-Token",
+    ]
+
+    def __init__(self, event: dict, methods: List[str], origins: List[str] = None):
+        super().__init__(event)
+
+        self._allowed_origins = origins if origins else env.get("CORS_ORIGIN", "").split(",")
+        self._check_for_wildcard()
```
Comment (songofroland) [inline lbz/resource.py:110]: Let's not mutate the state, rather set variables explicitly

How about:

```
    def __init__(self, event: dict, methods: List[str], origins: List[str] = None):
        super().__init__(event)

        allowed_origins = origins if origins else env.get("CORS_ORIGIN", "").split(",")
        self._allowed_origins = self._get_allowed_origin(allowed_origins)
        self._resp_headers = {
            ALLOW_ORIGIN_HEADER: self._allowed_origins[0],
            "Access-Control-Allow-Headers": ", ".join(self._cors_headers),
            "Access-Control-Allow-Methods": ", ".join([*methods, "OPTIONS"]),
        }

    def _get_allowed_origin(self, allowed_origins: List[str]) -> List[str]:
        actual_origin = self.request.headers.get("Origin")

        for allowed_origin in allowed_origins:
            if allowed_origin == actual_origin:
                return [actual_origin]
            elif "*" in allowed_origin:
                prefix, suffix = allowed_origin.split("*")
                if actual_origin.startswith(prefix) and actual_origin.endswith(suffix):
                    return [actual_origin]

        return allowed_origins
```

(it should be equivalent, not tested tho)

File: lbz/resource.py [hunk 2]
```diff
@@ -86,3 +91,86 @@ def pre_request_hook(self):
...
+            "Access-Control-Allow-Headers": ", ".join(self._cors_headers),
+            "Access-Control-Allow-Methods": ", ".join([*methods, "OPTIONS"]),
+        }
+
+    def __call__(self):
+        if self.method == "OPTIONS":
+            return Response("", headers=self.resp_headers(), status_code=HTTPStatus.NO_CONTENT)
+
+        resp = super().__call__()
+        if resp.status_code >= 400 and ALLOW_ORIGIN_HEADER not in resp.headers:
```
Comment (songofroland) [inline lbz/resource.py:122]: @grzegorzpro isn't it use case specific?

File: tests/test_resource.py [hunk 3]
```diff
@@ -210,3 +214,122 @@ def a(self):
             "message": "Server got itself in trouble",
             "request_id": ANY,
         }
+
+
+origin_localhost = "http://localhost:3000"
```
Comment (songofroland) [inline tests/test_resource.py:219]:
```
ORIGIN_LOCALHOST
ORIGIN_DEMO
```

File: tests/test_resource.py [hunk 4]
```diff
@@ -210,3 +214,122 @@ def a(self):
...
+        assert self.make_cors_handler().resp_headers_json["Content-Type"] == "application/json"
+
+    def test_resp_headers_no_content_type_by_default(self):
+        assert self.make_cors_handler().resp_headers().get("Content-Type") == None
+
+    def test_options_request(self):
+        inst = self.make_cors_handler(req_origin=origin_demo)
+        inst.method = "OPTIONS"
+        assert inst().headers == {
+            "Access-Control-Allow-Headers": ANY,
```
Comment (songofroland) [inline tests/test_resource.py:290]: why `ANY`?

File: CHANGELOG.md [hunk 5]
```diff
@@ -124,3 +124,9 @@ Released 2021-05-28
 Released 2021-06-13
 
 - Improve the messages that are logged in the entire library
+
+### Version 0.3.11
+Released 2021-06-30
+
+- Adds CORS-enabled Rosource class
```
Comment (songofroland) [inline CHANGELOG.md:131]: typo: `Resource`

File: CHANGELOG.md [hunk 6]
```diff
@@ -124,3 +124,9 @@ Released 2021-05-28
...
+- Adds Paginated Resource Helper class
```
Comment (songofroland) [inline CHANGELOG.md:132]: Add~~s~~

File: lbz/resource.py [hunk 7]
```diff
@@ -123,28 +125,23 @@ def __call__(self):
...
-                if org.startswith(serv) and org.endswith(domain):
-                    self._allowed_origins = [org]
-                    break
+    def _get_allowed_origins(self, origins):
+        if orig := self.request.headers.get("Origin"):
+            for aorg in origins:
+                if orig == aorg:
+                    return [orig]
+                if "*" in aorg:
+                    serv, domain = aorg.split("*")
```
Comment (songofroland) [inline lbz/resource.py:134]: I understand `aorg`, but `serv` instead of `server` does not look good

File: tests/test_resource.py [hunk 8]
```diff
@@ -230,50 +230,53 @@ def make_cors_handler(self, origins: List[str] = None, req_origin: str = None) -
...
     def test_cors_origin_headers_from_wildcard(self):
         assert (
             self.make_cors_handler(
                 origins=["https://*.lb.com"], req_origin="https://dev.test.lb.com"
-            ).resp_headers_json["Access-Control-Allow-Origin"]
+            ).resp_headers_json[ALLOW_ORIGIN_HEADER]
             == "https://dev.test.lb.com"
         )
 
+    def test_cors_origin_headers_from_wildcard_lambda(self):
```
Comment (songofroland) [inline tests/test_resource.py:268]: `test_cors_origin_headers_from_wildcard_lambda` -> `test_cors_origin_headers_from_wildcard_no_origin` imho

**Review by songofroland — APPROVED** [review]

File: lbz/resource.py (hunk 7 above)
Comment (redlickigrzegorz) [reply lbz/resource.py:134]: I don't think we should abbreviate variable names unless we have a good reason to do so - having the full names even for `aorg` and `orig` shouldn't be problematic (they are not too long too) 😉

Comment (pdyba) [reply lbz/resource.py:134]: orig is only becouse You have origins next to it and having origins and origin in same line is missleading.

**Review by songofroland — APPROVED** [review]

File: tests/test_resource.py [hunk 9]
```diff
@@ -210,3 +214,134 @@ def a(self):
...
+
+    def test_cors_origin_headers_from_env_are_correct_1(self):
+        assert self.make_cors_handler().resp_headers_json[ALLOW_ORIGIN_HEADER] == ORIGIN_LOCALHOST
+
+    def test_cors_origin_headers_from_env_are_correct_2(self):
+        assert (
+            self.make_cors_handler(req_origin=ORIGIN_EXAMPLE).resp_headers_json[
+                ALLOW_ORIGIN_HEADER
+            ]
+            == ORIGIN_EXAMPLE
```
Comment (redlickigrzegorz) [inline tests/test_resource.py:240]: Can we improve the formatting of all those tests? I think splitting into several lines will make them more readable 🙂

**Pull request merged**